pyro/infer/mcmc/api.py

```diff
@@ -0,0 +1,322 @@
+            while active_workers:
+                try:
+                    chain_id, val = self.result_queue.get(timeout=5)
+                    self.events[chain_id].set()
+                except queue.Empty:
+                    continue
+                if isinstance(val, Exception):
+                    # Exception trace is already logged by worker.
+                    raise val
+                if val is not None:
+                    yield val, chain_id
+                else:
+                    active_workers -= 1
+        finally:
+            self.terminate()
+
+
+class MCMC(object):
+    """
+    Wrapper class for Markov Chain Monte Carlo algorithms. Specific MCMC algorithms
+    are TraceKernel instances and need to be supplied as a ``kernel`` argument
+    to the constructor.
+
+    .. note:: The case of `num_chains > 1` uses python multiprocessing to
+        run parallel chains in multiple processes. This goes with the usual
+        caveats around multiprocessing in python, e.g. the model used to
+        initialize the ``kernel`` must be serializable via `pickle`, and the
+        performance / constraints will be platform dependent (e.g. only
+        the "spawn" context is available in Windows). This has also not
+        been extensively tested on the Windows platform.
+
+    :param kernel: An instance of the ``TraceKernel`` class, which when
+        given an execution trace returns another sample trace from the target
+        (posterior) distribution.
+    :param int num_samples: The number of samples that need to be generated,
+        excluding the samples discarded during the warmup phase.
+    :param int warmup_steps: Number of warmup iterations. The samples generated
+        during the warmup phase are discarded. If not provided, default is
+        half of `num_samples`.
+    :param int num_chains: Number of MCMC chains to run in parallel. Depending on
+        whether `num_chains` is 1 or more than 1, this class internally dispatches
+        to either `_SingleSampler` or `_ParallelSampler`.
+    :param dict initial_params: dict containing initial tensors to initiate
+        the markov chain. The leading dimension's size must match that of
+        `num_chains`. If not specified, parameter values will be sampled from
+        the prior.
+    :param hook_fn: Python callable that takes in `(kernel, samples, stage, i)`
+        as arguments. stage is either `sample` or `warmup` and i refers to the
+        i'th sample for the given stage. This can be
+    :param str mp_context: Multiprocessing context to use when `num_chains > 1`.
+        Only applicable for Python 3.5 and above. Use `mp_context="spawn"` for
+        CUDA.
+    :param bool disable_progbar: Disable progress bar and diagnostics update.
+    """
+    def __init__(self, kernel, num_samples, warmup_steps=None, initial_params=None,
+                 num_chains=1, hook_fn=None, mp_context=None, disable_progbar=False, transforms=None):
```

Comment:
nit: we need to add docs for `transforms`. But let's discuss a bit more about this point. I get it now that `mcmc` need to access `transforms`, which is not possible if they are created in child processes. I think that we can just call `initialize_model` here, rather than requires `transforms` to both HMC and MCMC, as you suggested. WDYT?

Reply:
If we use `potential_fn`, we shouldn't need to provide `transforms` to the HMC or NUTS kernel, and I think it makes more sense to provide transforms directly to the MCMC class instead. 

I will try to make it so that we do not need to provide transforms to both the classes though, that would be quite needless.